File: src/RunEvents.rs
// Import crates
extern crate ev3dev_lang_rust;


use log::{error, info};
use crate::consts::{LDRIVEENC, RDRIVEENC, LTOOLENC, RTOOLENC, CENTERBUTTON, DRIVEENC, LDRIVESPEED, RDRIVESPEED, LTOOLSPEED, RTOOLSPEED};
use std::time::{Instant};




// Local modules
use super::{
    COLOURSENS, DEBUG, GYRO, LDRIVECOR,
    RDRIVECOR,
};
use crate::Actuators::setMotorPow;
use crate::Events::{Event, FuncTypes};

use crate::ProcessLoop::SensorActuatorValues;
use crate::ReadSensors::getSensorValue;
use crate::PID::ComputePID;


fn MathFunc(inp: f32, func: &mut FuncTypes) -> f32 {
    match func {
        FuncTypes::ConstFunc { c } => *c,

        FuncTypes::LinearFunc {
            m,
            e,
            step_prev,
            lb,
            hb,
        } => {
            if *step_prev < 0.0 {
                *step_prev = inp;
            }

            let result: f32 = (inp - *step_prev) * (*m) + *e;
            if result > *hb {
                *hb
            } else if result < *lb {
                return *lb;
            } else {
                return (result * 50.0).round() / 50.0;
            }
        }

        FuncTypes::QuadFunc {
            a,
            b,
            c,
            step_prev,
            lb,
            hb,
        } => {
            if *step_prev == -1.0 {
                *step_prev = inp;
            }
            let x = inp - *step_prev;
            let result: f32 = (*a) * x.powf(2.0) + *b * x + *c;

            if result > *hb {
                *hb
            } else if result < *lb {
                return *lb;
            } else {
                return (result * 50.0).round() / 50.0;
            }
        }
    }
}

pub fn setReadSensor(sensor_id: i8, sensor_act_values: &mut SensorActuatorValues) 
{
    match sensor_id {
        LDRIVEENC => {
            sensor_act_values.lDriveMotorEncRead = true;
        }

        RDRIVEENC => {
            sensor_act_values.rDriveMotorEncRead = true;
        }

        LTOOLENC => {
            sensor_act_values.lToolMotorEncRead = true;
        }

        RTOOLENC => {
            sensor_act_values.rToolMotorEncRead = true;
        }

        LDRIVESPEED => {
            sensor_act_values.lDriveMotorSpeedRead = true;
        }

        RDRIVESPEED => {
            sensor_act_values.rDriveMotorSpeedRead = true;
        }

        LTOOLSPEED => {
            sensor_act_values.lToolMotorSpeedRead = true;
        }

        RTOOLSPEED => {
            sensor_act_values.rToolMotorSpeedRead = true;
        }

        GYRO => {
            sensor_act_values.gyroRead = true;
        }

        CENTERBUTTON => {
            sensor_act_values.centerButtonRead = true;
        }

        DRIVEENC => {
            sensor_act_values.lDriveMotorEncRead = true;
            sensor_act_values.rDriveMotorEncRead = true;
        }

        _ => {}
    }
}

pub fn RunEvents(
    event_list: &mut Vec<Event>,
    ActiveTable: &[bool],
    CondTable: &mut [bool],
    sensor_act_values: &mut SensorActuatorValues,
    sys_time: &Instant,
    running: &mut bool
) {
    sensor_act_values.currentTime = sys_time.elapsed().as_secs_f32();
    
    for _event in event_list {
        match _event {

            Event::SensorValue {
                event,
                sensor_target,
                sensor_prev,
                sensor_id,
                expr,
                sensvalcondid,
            } => {
                if ActiveTable[event.process_id] {
                    let sensor_value: f32 = getSensorValue(*sensor_id, sensor_act_values);
                    if *sensor_prev < -9998.0 {
                        *sensor_prev = sensor_value;
                    }

                    match expr {
                        '>' => {
                            CondTable[*sensvalcondid] =
                                sensor_value - *sensor_prev >= *sensor_target
                        }
                        '<' => {
                            CondTable[*sensvalcondid] =
                                sensor_value - *sensor_prev <= *sensor_target
                        }
                        _ => {
                            if DEBUG {
                                error!("Invalid character {} at Events::SensorValue", expr);
                            }
                        }
                    }

                    setReadSensor(*sensor_id, sensor_act_values);
                }
            }

            // Motor Control
            Event::MotorSpeedControl {
                event,
                motor_id,
                sensor_id,
                func,
            } => {
                if ActiveTable[event.process_id] {
                    setMotorPow(
                        MathFunc(getSensorValue(*sensor_id, sensor_act_values), func),
                        *motor_id,
                        sensor_act_values,
                    );
                }
            }

            // PID Control
            Event::PIDGyro {
                event,
                heading,
                pid,
                motor_correction,
                sensor_prev,
            } => {
                if ActiveTable[event.process_id] {
                    let sensor_value: f32 = getSensorValue(GYRO, sensor_act_values);
                    if *sensor_prev < -9998.0 {
                        *sensor_prev = sensor_value;
                    }

                    *motor_correction = ComputePID(sensor_value - *sensor_prev, heading, pid);
                    setMotorPow(-*motor_correction, LDRIVECOR, sensor_act_values);
                    setMotorPow(*motor_correction, RDRIVECOR, sensor_act_values);

                    sensor_act_values.gyroRead = true;
                }
            }

            Event::PIDLine {
                event,
                brightness_target,
                pid,
                motor_correction,
            } => {
                if ActiveTable[event.process_id] {
                    *motor_correction = ComputePID(
                        getSensorValue(COLOURSENS, sensor_act_values),
                        brightness_target,
                        pid,
                    );
                    setMotorPow(*motor_correction, LDRIVECOR, sensor_act_values);
                    setMotorPow(-*motor_correction, RDRIVECOR, sensor_act_values);
                }
            }

            Event::PIDHold { event, pid: _, motor_correction: _} => if ActiveTable[event.process_id] {},

            // Compute and Timer
            Event::ComputeMotorStall {
                event,
                min_mov_avg_speed,
                buffer,
                buffer_size,
                motor_id,
            } => {
                if ActiveTable[event.process_id] {
                    buffer.push_back(getSensorValue(*motor_id, sensor_act_values));
                    
                    if buffer.len() == *buffer_size {
                        let avg: f32 = buffer.iter().sum::<f32>() / *buffer_size as f32;
                        CondTable[event.term_conditions_id] = avg < *min_mov_avg_speed;

                        buffer.pop_front();
                    }
                    setReadSensor(*motor_id, sensor_act_values);
                }
            },

            Event::Timer {
                event,
                time,
                time_prev,
            } => {
                if ActiveTable[event.process_id] {
                    if *time_prev == -1.0 {
                        *time_prev = sys_time.elapsed().as_secs_f32();
                    } else {
                        let time_passed = sys_time.elapsed().as_secs_f32() - *time_prev;

                        if time_passed >= *time {
                            CondTable[event.term_conditions_id] = true;
                        }
                    }
                }
            }

            Event::HaltProcessLoop { 
                event 
            } => {
                if ActiveTable[event.process_id] {
                    *running = false;
                }
            }
            
            _ => {}
        }
    }
}

File: spike-prime-pybricks/RunEvents.py
from consts import *
from PID import compute_pid
from ReadSensors import get_sensor_value
from Actuators import set_motor_pow


#import micropython
import time

#@micropython.native
def math_func(x, func):
    if func.t == FUNC_CONST:
        return func.c
    
    elif func.t == FUNC_LINEAR:
        if func.step_prev < 0:
            func.step_prev = x

        result = func.m * (x - func.step_prev) + func.e
        if result > func.hb:
            return func.hb
        elif result < func.lb:
            return func.lb
        else:
            return round(result*50)/50
        
    elif func.t == FUNC_QUADRATIC:
        return func.a * x * x + func.b * x + func.c

#@micropython.native
def set_read_sensor(sensor_id, sensor_act_values):
    # Logic to set sensor read flags
    if sensor_id == LDRIVEENC:
        sensor_act_values.lDriveMotorEncRead = True
    elif sensor_id == RDRIVEENC:
        sensor_act_values.rDriveMotorEncRead = True
    elif sensor_id == LTOOLENC:
        sensor_act_values.lToolMotorEncRead = True
    elif sensor_id == RTOOLENC:
        sensor_act_values.rToolMotorEncRead = True
    elif sensor_id == LDRIVESPEED:
        sensor_act_values.lDriveMotorSpeedRead = True
    elif sensor_id == RDRIVESPEED:
        sensor_act_values.rDriveMotorSpeedRead = True
    elif sensor_id == LTOOLSPEED:
        sensor_act_values.lToolMotorSpeedRead = True
    elif sensor_id == RTOOLSPEED:
        sensor_act_values.rToolMotorSpeedRead = True
    elif sensor_id == DRIVESPEED:
        sensor_act_values.lDriveMotorSpeedRead = True
        sensor_act_values.rDriveMotorSpeedRead = True
    elif sensor_id == GYRO:
        sensor_act_values.gyroRead = True
    elif sensor_id == RIGHTBUTTON:
        sensor_act_values.rightButtonRead = True
    elif sensor_id == DRIVEENC:
        sensor_act_values.lDriveMotorEncRead = True
        sensor_act_values.rDriveMotorEncRead = True

#@micropython.native
def run_events(event_list, active_table, cond_table, sensor_act_values):
    sensor_act_values.currentTime = time.time()
    running = True
    for event in event_list:
        event_type = event.t

        if event_type == EVENT_SENSORVALUE and active_table[event.event.process_id]:
            sensor_value = get_sensor_value(event.sensor_id, sensor_act_values)
            if event.sensor_prev < -9998.0:
                event.sensor_prev = sensor_value

            if event.expr == '>':
                cond_table[event.sensvalcondid] = (sensor_value - event.sensor_prev >= event.sensor_target)
            elif event.expr == '<':
                cond_table[event.sensvalcondid] = (sensor_value - event.sensor_prev <= event.sensor_target)
            else:
                if DEBUG:
                    print(f"Invalid character {event.expr} at Events::SensorValue")
            
            set_read_sensor(event.sensor_id, sensor_act_values)

        elif event_type == EVENT_MOTORSPEEDCONTROL and active_table[event.event.process_id]:
            motor_value = math_func(get_sensor_value(event.sensor_id, sensor_act_values), event.func)
            set_motor_pow(motor_value, event.motor_id, sensor_act_values)
            set_read_sensor(event.sensor_id, sensor_act_values)

        elif event_type == EVENT_PIDGYRO and active_table[event.event.process_id]:
            sensor_value = get_sensor_value(GYRO, sensor_act_values)
            if event.sensor_prev < -9998.0:
                event.sensor_prev = sensor_value

            event.motor_correction = compute_pid(
                sensor_value - event.sensor_prev,
                event.heading,
                sensor_act_values.currentTime - sensor_act_values.timePrev,
                event.pid
            )

            set_motor_pow(-event.motor_correction, LDRIVECOR, sensor_act_values)
            set_motor_pow(event.motor_correction, RDRIVECOR, sensor_act_values)
            sensor_act_values.gyroRead = True
            set_read_sensor(DRIVEENC, sensor_act_values)

        elif event_type == EVENT_PIDLINE and active_table[event.event.process_id]:
            event.motor_correction = compute_pid(
                get_sensor_value(COLOURSENS, sensor_act_values),
                event.brightness_target,
                sensor_act_values.currentTime - sensor_act_values.timePrev,
                event.pid
            )
            set_motor_pow(event.motor_correction, LDRIVECOR, sensor_act_values)
            set_motor_pow(-event.motor_correction, RDRIVECOR, sensor_act_values)

        elif event_type == EVENT_COMPUTEMOTORSTALL and active_table[event.event.process_id]:
            event.buffer.append(get_sensor_value(event.motor_id, sensor_act_values))
            if len(event.buffer) == event.buffer_size:
                avg_speed = sum(event.buffer) / len(event.buffer)
                cond_table[event.event.term_conditions_id] = (avg_speed < event.min_mov_avg_speed)
                event.buffer.pop(0)
            set_read_sensor(event.motor_id, sensor_act_values)

        elif event_type == EVENT_TIMER and active_table[event.event.process_id]:
            if event.time_prev == -1.0:
                event.time_prev = sensor_act_values.currentTime
            else:
                time_passed = sensor_act_values.currentTime - event.time_prev
                if time_passed >= event.time:
                    cond_table[event.event.term_conditions_id] = True
        
        
        elif event_type == EVENT_HALT and active_table[event.event.process_id]:
            running = False
            
    return running
